File: stratosphere-core/src/main/java/eu/stratosphere/api/functions/Function.java
/***********************************************************************************************************************
 * Copyright (C) 2010-2013 by the Stratosphere project (http://stratosphere.eu)
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 **********************************************************************************************************************/

package eu.stratosphere.api.functions;

import eu.stratosphere.configuration.Configuration;

/**
 * Abstract stub class for all PACT stubs. PACT stubs must be overwritten to provide user implementations
 * for PACT programs.
 */
public interface Function {
	
	/**
	 * Initialization method for the stub. It is called before the actual working methods 
	 * (like <i>map</i> or <i>match</i>). This method should be used for configuration and 
	 * initialization of the stub implementation.
	 * <p>
	 * This method receives the parameters attached to the contract. Consider the following pseudo code example,
	 * which realizes a parameterizable filter:
	 * <code>
	 * public Plan getPlan(String... args)
	 * {
	 *     MapContract mc = new MapContract(MyMapper.class, "My Mapper");
	 *     mc.setDegreeOfParallelism(48);
	 *     mc.getStubParameters().setString("foo", "bar");
	 *     
	 *      ...
	 *      
	 *      Plan plan = new Plan(...);
	 *      return plan;
	 * }
	 * 
	 * public class MyMapper extends MapStub {
	 * 
	 *     private String searchString;
	 *     
	 *     public void open(Configuration parameters) {
	 *         this.searchString = parameters.getString("foo", null);
	 *         // searchString will be "bar" when job is started
	 *     }
	 *     
	 *     public void map(PactRecord record, Collector collector) {
	 *         if ( record.getValue(0, PactString.class).equals(this.searchString) ) {
	 *             collector.emit(record);
	 *         }
	 *     }
	 * }
	 * </code>
	 * <p>
	 * By default, this method does nothing.
	 * 
	 * @param parameters The configuration containing the parameters attached to the contract. 
	 * 
	 * @throws Exception Implementations may forward exceptions, which are caught by the runtime. When the
	 *                   runtime catches an exception, it aborts the task and lets the fail-over logic
	 *                   decide whether to retry the task execution.
	 * 
	 * @see eu.stratosphere.configuration.Configuration
	 */
	void open(Configuration parameters) throws Exception;

	/**
	 * Teardown method for the user code. It is called after the last call to the main working methods
	 * (e.g. <i>map</i> or <i>match</i>). It is called also when the task is aborted, in which case exceptions
	 * thrown by this method are logged but ignored. 
	 * <p>
	 * This method should be used for clean up.
	 * 
	 * @throws Exception Implementations may forward exceptions, which are caught by the runtime. When the
	 *                   runtime catches an exception, it aborts the task and lets the fail-over logic
	 *                   decide whether to retry the task execution.
	 */
	void close() throws Exception;
	
	
	/**
	 * Gets the context that contains information about the UDF's runtime.
	 * 
	 * @return The UDF's runtime context.
	 */
	RuntimeContext getRuntimeContext();
	
	/**
	 * Sets the stub's runtime context. Called by the framework when creating a parallel instance of the stub.
	 *  
	 * @param t The runtime context.
	 */
	void setRuntimeContext(RuntimeContext t);
}

File: flink-libraries/flink-python/src/main/python/org/apache/flink/python/api/flink/functions/Function.py
# ###############################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
from abc import ABCMeta, abstractmethod
from collections import deque
from flink.connection import Connection, Iterator, Collector
from flink.connection.Iterator import IntegerDeserializer, StringDeserializer, _get_deserializer
from flink.functions import RuntimeContext


class Function(object):
    __metaclass__ = ABCMeta

    def __init__(self):
        self._connection = None
        self._iterator = None
        self._collector = None
        self.context = None
        self._env = None

    def _configure(self, input_file, output_file, mmap_size, port, env, info, subtask_index):
        self._connection = Connection.BufferingTCPMappedFileConnection(input_file, output_file, mmap_size, port)
        self._iterator = Iterator.Iterator(self._connection, env)
        self._collector = Collector.Collector(self._connection, env, info)
        self.context = RuntimeContext.RuntimeContext(self._iterator, self._collector, subtask_index)
        self._env = env
        if info.chained_info is not None:
            info.chained_info.operator._configure_chain(self.context, self._collector, info.chained_info)
            self._collector = info.chained_info.operator

    def _configure_chain(self, context, collector, info):
        self.context = context
        if info.chained_info is None:
            self._collector = collector
        else:
            self._collector = info.chained_info.operator
            info.chained_info.operator._configure_chain(context, collector, info.chained_info)

    @abstractmethod
    def _run(self):
        pass

    def _close(self):
        self._collector._close()
        if self._connection is not None:
            self._connection.close()

    def _go(self):
        self._receive_broadcast_variables()
        self._run()

    def _receive_broadcast_variables(self):
        con = self._connection
        deserializer_int = IntegerDeserializer()
        broadcast_count = deserializer_int.deserialize(con.read_secondary)
        deserializer_string = StringDeserializer()
        for _ in range(broadcast_count):
            name = deserializer_string.deserialize(con.read_secondary)
            bc = deque()
            if con.read_secondary(1) == b"\x01":
                serializer_data = _get_deserializer(con.read_secondary, self._env._types)
                value = serializer_data.deserialize(con.read_secondary)
                bc.append(value)
                while con.read_secondary(1) == b"\x01":
                    con.read_secondary(serializer_data.get_type_info_size()) #skip type info
                    value = serializer_data.deserialize(con.read_secondary)
                    bc.append(value)
            self.context._add_broadcast_variable(name, bc)



